### ISSUE TITLE
Increase memory causes bosh agent fails to start

### ISSUE THREAD

**Issue by edwardstudy** [cloudfoundry/bosh-agent#184]
Hi, we are doing test for softlayer-cpi with xenial stemcell(agent v2.123.0). 

We just increased cell memory from 32G to 48G which caused the ephemeral disk was re-patitioned and formatted. Trusty stemcell doesn't have this problem.

For Xenial stemcell, after os reload, the memory is successfully modified (seen from SL portal) but bosh agent failed to start:

```
2018-08-01_10:02:27.42200 [linuxPlatform] 2018/08/01 10:02:27 INFO - Partitioning `/dev/xvdc' with [[Type: swap, SizeInBytes: 50636873728] [Type: linux, SizeInBytes: 271485673472]]
2018-08-01_10:02:27.42203 [Cmd Runner] 2018/08/01 10:02:27 DEBUG - Running command 'parted -m /dev/xvdc unit B print'
2018-08-01_10:02:27.43166 [Cmd Runner] 2018/08/01 10:02:27 DEBUG - Stdout: BYT;
2018-08-01_10:02:27.43167 /dev/xvdc:322122547200B:xvd:512:512:gpt:Xen Virtual Block Device:;
2018-08-01_10:02:27.43168 1:1048576B:33726398463B:33725349888B:linux-swap(v1):bosh-partition-0:;
2018-08-01_10:02:27.43168 2:33726398464B:322121498623B:288395100160B:ext4:bosh-partition-1:;
2018-08-01_10:02:27.43172 [Cmd Runner] 2018/08/01 10:02:27 DEBUG - Stderr:
2018-08-01_10:02:27.43173 [Cmd Runner] 2018/08/01 10:02:27 DEBUG - Successful: true (0)
2018-08-01_10:02:27.43175 [main] 2018/08/01 10:02:27 ERROR - App setup Running bootstrap: Setting up ephemeral disk: Partitioning ephemeral disk: Partitioning ephemeral disk '/dev/xvdc': '/dev/xvdc' contains a partition created by bosh. No partitioning is allowed.
2018-08-01_10:02:27.43176 [main] 2018/08/01 10:02:27 ERROR - Agent exited with error: Running bootstrap: Setting up ephemeral disk: Partitioning ephemeral disk: Partitioning ephemeral disk '/dev/xvdc': '/dev/xvdc' contains a partition created by bosh. No partitioning is allowed.
```

I saw a [story](https://www.pivotaltracker.com/story/show/141166603) about parted re-partition stuffs. I'm still wondering why. Thanks.

**Comment by dpb587-pivotal** [comment]
Did the ephemeral disk change as part of the cell memory change? Was the VM previously running Trusty?

You are correct, that story is probably where this change came in. As mentioned in the description, it was a defensive change to prevent data loss (primarily intended for the persistent disk).

But, I thought we did not try and repartition a disk if it was already partitioned and did not need to change. That may be different in the ephemeral case, and especially in the OS reload case since we do not use or specifically test that scenario.

In Trusty, what was the actual behavior when it did not error? For an exactly equivalent change with Trusty, can you confirm from logs if it still reformatted/repartitioned the ephemeral disk?

**Comment by edwardstudy** [comment]
@dpb587-pivotal Hi, Danny. The VM is created originally from Xenial. We just reloaded it to change memory.

IIRC, agent re-partitioned ephemeral disk when changing memory in Trusty.

**Comment by edwardstudy** [comment]
I tested on the same VM(4G -> 8G) for skip bosh-partition-* label scanning. The agent still failed to partition:

```
2018-08-02_14:38:40.24420 [Cmd Runner] 2018/08/02 14:38:40 DEBUG - Running command 'parted -m /dev/xvdc unit B print'
2018-08-02_14:38:40.26077 [Cmd Runner] 2018/08/02 14:38:40 DEBUG - Stdout: BYT;
2018-08-02_14:38:40.26082 /dev/xvdc:107374182400B:xvd:512:512:gpt:Xen Virtual Block Device:;
2018-08-02_14:38:40.26092 1:1048576B:4131389439B:4130340864B:linux-swap(v1):bosh-partition-0:;
2018-08-02_14:38:40.26095 2:4131389440B:107373133823B:103241744384B:ext4:bosh-partition-1:;
2018-08-02_14:38:40.26105 [Cmd Runner] 2018/08/02 14:38:40 DEBUG - Stderr:
2018-08-02_14:38:40.26115 [Cmd Runner] 2018/08/02 14:38:40 DEBUG - Successful: true (0)
2018-08-02_14:38:40.26125 [attemptRetryStrategy] 2018/08/02 14:38:40 DEBUG - Making attempt #0
2018-08-02_14:38:40.26134 [Cmd Runner] 2018/08/02 14:38:40 DEBUG - Running command 'parted -s /dev/xvdc unit B mkpart bosh-partition-0 1048576 8359247871'
2018-08-02_14:38:40.27055 [Cmd Runner] 2018/08/02 14:38:40 DEBUG - Stdout:
2018-08-02_14:38:40.27061 [Cmd Runner] 2018/08/02 14:38:40 DEBUG - Stderr: Error: You requested a partition from 1048576B to 8359247360B (sectors 2048..16326655).
2018-08-02_14:38:40.27069 The closest location we can manage is 1048064B to 1048064B (sectors 2047..2047).
2018-08-02_14:38:40.27072 [Cmd Runner] 2018/08/02 14:38:40 DEBUG - Successful: false (1)
2018-08-02_14:38:40.27079 [PartedPartitioner] 2018/08/02 14:38:40 ERROR - Failed with an error: Running command: 'parted -s /dev/xvdc unit B mkpart bosh-partition-0 1048576 8359247871', stdout: '', stderr: 'Error: You requested a partition from 1048576B to 8359247360B (sectors 2048..16326655).
2018-08-02_14:38:40.27083 The closest location we can manage is 1048064B to 1048064B (sectors 2047..2047).
2018-08-02_14:38:40.27090 ': exit status 1
```

**Comment by edwardstudy** [comment]
And agent on Trusty uses `sfdisk` to partition and re-partition:

1. partition:
```
2018-08-03_02:39:44.83236 [linuxPlatform] 2018/08/03 02:39:44 INFO - Creating swap & ephemeral partitions on ephemeral disk...
2018-08-03_02:39:44.83237 [linuxPlatform] 2018/08/03 02:39:44 DEBUG - Getting device size of `/dev/xvdc'
2018-08-03_02:39:44.83241 [Cmd Runner] 2018/08/03 02:39:44 DEBUG - Running command 'sfdisk -s /dev/xvdc'
2018-08-03_02:39:44.83464 [Cmd Runner] 2018/08/03 02:39:44 DEBUG - Stdout: 104857600
2018-08-03_02:39:44.83465 [Cmd Runner] 2018/08/03 02:39:44 DEBUG - Stderr:
2018-08-03_02:39:44.83466 [Cmd Runner] 2018/08/03 02:39:44 DEBUG - Successful: true (0)
2018-08-03_02:39:44.83466 [linuxPlatform] 2018/08/03 02:39:44 DEBUG - Calculating partition sizes of `/dev/xvdc', with available size 107374182400B
2018-08-03_02:39:44.83488 [linuxPlatform] 2018/08/03 02:39:44 INFO - Partitioning `/dev/xvdc' with [[Type: swap, SizeInBytes: 4137791488] [Type: linux, SizeInBytes: 103236390912]]
2018-08-03_02:39:44.83489 [Cmd Runner] 2018/08/03 02:39:44 DEBUG - Running command 'sfdisk -d /dev/xvdc'
2018-08-03_02:39:44.83650 [Cmd Runner] 2018/08/03 02:39:44 DEBUG - Stdout: # partition table of /dev/xvdc
2018-08-03_02:39:44.83651 unit: sectors
2018-08-03_02:39:44.83651
2018-08-03_02:39:44.83651 /dev/xvdc1 : start=        1, size=  8096759, Id=82
2018-08-03_02:39:44.83652 /dev/xvdc2 : start=  8096760, size=201615750, Id=83
2018-08-03_02:39:44.83652 /dev/xvdc3 : start=        0, size=        0, Id= 0
2018-08-03_02:39:44.83653 /dev/xvdc4 : start=        0, size=        0, Id= 0
2018-08-03_02:39:44.83653 [Cmd Runner] 2018/08/03 02:39:44 DEBUG - Stderr:
2018-08-03_02:39:44.83653 [Cmd Runner] 2018/08/03 02:39:44 DEBUG - Successful: true (0)
2018-08-03_02:39:44.83654 [Cmd Runner] 2018/08/03 02:39:44 DEBUG - Running command 'sfdisk -s /dev/xvdc1'
2018-08-03_02:39:44.83750 [Cmd Runner] 2018/08/03 02:39:44 DEBUG - Stdout: 4048379
2018-08-03_02:39:44.83753 [Cmd Runner] 2018/08/03 02:39:44 DEBUG - Stderr:
2018-08-03_02:39:44.83753 [Cmd Runner] 2018/08/03 02:39:44 DEBUG - Successful: true (0)
2018-08-03_02:39:44.83757 [Cmd Runner] 2018/08/03 02:39:44 DEBUG - Running command 'sfdisk -s /dev/xvdc2'
2018-08-03_02:39:44.83867 [Cmd Runner] 2018/08/03 02:39:44 DEBUG - Stdout: 100807875
2018-08-03_02:39:44.83868 [Cmd Runner] 2018/08/03 02:39:44 DEBUG - Stderr:
2018-08-03_02:39:44.83868 [Cmd Runner] 2018/08/03 02:39:44 DEBUG - Successful: true (0)
2018-08-03_02:39:44.83880 [Cmd Runner] 2018/08/03 02:39:44 DEBUG - Running command 'sfdisk -s /dev/xvdc'
2018-08-03_02:39:44.83965 [Cmd Runner] 2018/08/03 02:39:44 DEBUG - Stdout: 104857600
2018-08-03_02:39:44.83966 [Cmd Runner] 2018/08/03 02:39:44 DEBUG - Stderr:
2018-08-03_02:39:44.83966 [Cmd Runner] 2018/08/03 02:39:44 DEBUG - Successful: true (0)
2018-08-03_02:39:44.83967 [SfdiskPartitioner] 2018/08/03 02:39:44 INFO - /dev/xvdc already partitioned as expected, skipping
```

2. re-partition:
```
2018-08-03_02:54:22.22493 [linuxPlatform] 2018/08/03 02:54:22 DEBUG - Calculating partition sizes of `/dev/xvdc', with available size 107374182400B
2018-08-03_02:54:22.22493 [linuxPlatform] 2018/08/03 02:54:22 INFO - Partitioning `/dev/xvdc' with [[Type: swap, SizeInBytes: 8365645824] [Type: linux, SizeInBytes: 99008536576]]
2018-08-03_02:54:22.22494 [Cmd Runner] 2018/08/03 02:54:22 DEBUG - Running command 'sfdisk -d /dev/xvdc'
2018-08-03_02:54:22.22649 [Cmd Runner] 2018/08/03 02:54:22 DEBUG - Stdout: # partition table of /dev/xvdc
2018-08-03_02:54:22.22651 unit: sectors
2018-08-03_02:54:22.22651
2018-08-03_02:54:22.22652 /dev/xvdc1 : start=        1, size=  8096759, Id=82
2018-08-03_02:54:22.22652 /dev/xvdc2 : start=  8096760, size=201615750, Id=83
2018-08-03_02:54:22.22653 /dev/xvdc3 : start=        0, size=        0, Id= 0
2018-08-03_02:54:22.22653 /dev/xvdc4 : start=        0, size=        0, Id= 0
2018-08-03_02:54:22.22653 [Cmd Runner] 2018/08/03 02:54:22 DEBUG - Stderr:
2018-08-03_02:54:22.22654 [Cmd Runner] 2018/08/03 02:54:22 DEBUG - Successful: true (0)
2018-08-03_02:54:22.22654 [Cmd Runner] 2018/08/03 02:54:22 DEBUG - Running command 'sfdisk -s /dev/xvdc1'
2018-08-03_02:54:22.22756 [Cmd Runner] 2018/08/03 02:54:22 DEBUG - Stdout: 4048379
2018-08-03_02:54:22.22763 [Cmd Runner] 2018/08/03 02:54:22 DEBUG - Stderr:
2018-08-03_02:54:22.22765 [Cmd Runner] 2018/08/03 02:54:22 DEBUG - Successful: true (0)
2018-08-03_02:54:22.22775 [Cmd Runner] 2018/08/03 02:54:22 DEBUG - Running command 'sfdisk -s /dev/xvdc2'
2018-08-03_02:54:22.22873 [Cmd Runner] 2018/08/03 02:54:22 DEBUG - Stdout: 100807875
2018-08-03_02:54:22.22877 [Cmd Runner] 2018/08/03 02:54:22 DEBUG - Stderr:
2018-08-03_02:54:22.22879 [Cmd Runner] 2018/08/03 02:54:22 DEBUG - Successful: true (0)
2018-08-03_02:54:22.22887 [Cmd Runner] 2018/08/03 02:54:22 DEBUG - Running command 'sfdisk -s /dev/xvdc'
2018-08-03_02:54:22.22989 [Cmd Runner] 2018/08/03 02:54:22 DEBUG - Stdout: 104857600
2018-08-03_02:54:22.23000 [Cmd Runner] 2018/08/03 02:54:22 DEBUG - Stderr:
2018-08-03_02:54:22.23005 [Cmd Runner] 2018/08/03 02:54:22 DEBUG - Successful: true (0)
2018-08-03_02:54:22.23013 [attemptRetryStrategy] 2018/08/03 02:54:22 DEBUG - Making attempt #0
2018-08-03_02:54:22.23019 [Cmd Runner] 2018/08/03 02:54:22 DEBUG - Running command 'sfdisk -uM /dev/xvdc'
2018-08-03_02:54:22.24185 [Cmd Runner] 2018/08/03 02:54:22 DEBUG - Stdout:
2018-08-03_02:54:22.24194 Disk /dev/xvdc: 13054 cylinders, 255 heads, 63 sectors/track
2018-08-03_02:54:22.24199 Old situation:
2018-08-03_02:54:22.24203 Units = mebibytes of 1048576 bytes, blocks of 1024 bytes, counting from 0
2018-08-03_02:54:22.24208
2018-08-03_02:54:22.24213    Device Boot Start   End    MiB    #blocks   Id  System
2018-08-03_02:54:22.24218 /dev/xvdc1         0+  3953-  3954-   4048379+  82  Linux swap / Solaris
2018-08-03_02:54:22.24222 /dev/xvdc2      3953+ 102398- 98446- 100807875   83  Linux
2018-08-03_02:54:22.24227 /dev/xvdc3         0      -      0          0    0  Empty
2018-08-03_02:54:22.24232 /dev/xvdc4         0      -      0          0    0  Empty
2018-08-03_02:54:22.24237 New situation:
2018-08-03_02:54:22.24241 Units = mebibytes of 1048576 bytes, blocks of 1024 bytes, counting from 0
2018-08-03_02:54:22.24249
2018-08-03_02:54:22.24257    Device Boot Start   End    MiB    #blocks   Id  System
2018-08-03_02:54:22.24263 /dev/xvdc1         0+  7985-  7986-   8177084+  82  Linux swap / Solaris
2018-08-03_02:54:22.24269 /dev/xvdc2      7985+ 102398- 94414-  96679170   83  Linux
2018-08-03_02:54:22.24276 /dev/xvdc3         0      -      0          0    0  Empty
2018-08-03_02:54:22.24281 /dev/xvdc4         0      -      0          0    0  Empty
2018-08-03_02:54:22.24286 Successfully wrote the new partition table
2018-08-03_02:54:22.24291
2018-08-03_02:54:22.24296 Re-reading the partition table ...
2018-08-03_02:54:22.24302
2018-08-03_02:54:22.24306 [Cmd Runner] 2018/08/03 02:54:22 DEBUG - Stderr: Checking that no-one is using this disk right now ...
2018-08-03_02:54:22.24314 OK
2018-08-03_02:54:22.24318 Warning: no primary partition is marked bootable (active)
2018-08-03_02:54:22.24324 This does not matter for LILO, but the DOS MBR will not boot this disk.
2018-08-03_02:54:22.24329 If you created or changed a DOS partition, /dev/foo7, say, then use dd(1)
2018-08-03_02:54:22.24333 to zero the first 512 bytes:  dd if=/dev/zero of=/dev/foo7 bs=512 count=1
2018-08-03_02:54:22.24338 (See fdisk(8).)
2018-08-03_02:54:22.24343 [Cmd Runner] 2018/08/03 02:54:22 DEBUG - Successful: true (0)
2018-08-03_02:54:22.24348 [SfdiskPartitioner] 2018/08/03 02:54:22 INFO - Succeeded in partitioning /dev/xvdc with ,7978,S
2018-08-03_02:54:22.24355 ,,L
```

**Comment by dpb587-pivotal** [comment]
Recapping from our recent discussion...

 * any solution should consider what happens when the VM is rebooted; we should not erase the disk in a generic restart scenario
 * asked to look into OS reload hooks as alternatives to making changes to the generic partitioner; e.g. perhaps the OS reload could reset the partition table to make it appear empty at start

**Comment by mattcui** [comment]
Hi Danny @dpb587-pivotal , I would like to propose an approach to distinguish between os reload and agent restart, we could set a marker file like `ephemeral_disk_partitioned` under `/var/vcap/bosh` in the stemcell, the default value is `false`, which means the ephemeral disk needs to be partitioned. The agent will read in this file at the beginning of agent start, if it's `false`, the agent will partition the ephemeral disk and change to `true` in  `ephemeral_disk_partitioned`. So the ephemeral disk will not be re-partitioned for normal agent restart, but for new VM creation or os reload, the value is `false`, agent will partition the disk. What do you think? Thanks.

**Comment by mattcui** [comment]
@jfmyers9 One thing we need to consider is that the ephemeral disk should be cleaned up after OS reload, SoftLayer can't do that through the API call, this is why I proposed a solution in https://github.com/cloudfoundry/bosh-agent/issues/184#issuecomment-412249448, actually, we had a solution in existing agent code ([ScrubEphemeralDisk](https://github.com/cloudfoundry/bosh-agent/blob/master/platform/linux_platform.go#L59)), but it has some limitation, the ephemeral disk can't be cleaned up if the stemcell version is not changed (like `bosh recreate` command) after OS reload. So we hope to take this opportunity to fix the problem completely. Thanks.

**Comment by edwardstudy** [comment]
@jfmyers9 Thanks. As @mattcui pointed, the ephemeral disk didn't be formatted after OS reload, because agent skipped formatting old ephemeral disk when old signatures have still existed.

So agent should also format old ephemeral disk after repartition.

**Comment by jfmyers9** [comment]
Hey @edwardstudy and @mattcui,

I think the proposed `EphemeralDevicePartioner` would be able to encapsulate that logic as well. You could imagine it's work flow being something like:
* Check to see if existing partitions match desired partitions. Exit if there is nothing to do.
* Remove existing partitions as this is an ephemeral disk.
* Format the disk
* Partition with desired partitions.

Do you have any concerns with this proposed solution? If not, we should close https://github.com/cloudfoundry/bosh-agent/pull/186 and could you submit a new PR implementing this idea? If not, we also will create a story to track this issue, and we may be able to get to it at some point depending on prioritization.

Best,

@jfmyers9 && @jrussett, CF BOSH

**Comment by dpb587-pivotal** [comment]
@jfmyers9 @jrussett does the proposed workflow cover the scenario of a plain `bosh recreate` which should result in an empty ephemeral disk at some point during startup?

**Comment by jfmyers9** [comment]
We updated the strategy to reflect an empty ephemeral disk. It would look something like:

* if there are existing partitions
  * exit if existing partitions match desired partitions
  * remove existing partitions
* partition with desired partitions

The linux platform already formats the ephemeral disk's partitions if they are not currently formatted as `ext4` or `xfs`. In the case of a used ephemeral disk, removing the partitions and recreating them should trigger a reformat. In the case of an empty ephemeral disk, it will be partitioned and formatted. In the case of an ephemeral disk that has the correct partitions, nothing will happen.

Does that seem correct?

**Comment by dpb587-pivotal** [comment]
That seems correct...

But I'm not seeing how it would cause a disk to be erased in the scenario where you do a `bosh recreate`. The specs are exactly the same, but, in OS reload case, the VM boots with the existing ephemeral disk which matches the desired partition sizes, so the proposed workflow doesn't do anything to it. Which then means it retains the existing ephemeral data; which seems bad.

Let me know if I'm missing something though; then should chat more irl.

**Comment by edwardstudy** [comment]
@dpb587-pivotal Hi, Danny. How about saving `ephemeral_disk_partitioned` state and doing ephemeral disk partition if `ephemeral_disk_partitioned` is false and updating state. You can see details from: https://github.com/cloudfoundry/bosh-agent/pull/186

If @jfmyers9's strategy is confirmed, I could update code and submit a new PR.

**Comment by dpb587-pivotal** [comment]
cc @jfmyers9 do you have an opinion on the `ephemeral_disk_partitioned` state file?

That sort of file feels very SoftLayer-specific requirement to me; it's not a standard requirement in any other IaaSes, so would hope we can continue brainstorming on more generic solutions, too. Conceptually, it makes sense why OS reload would need it... but any other ideas for how to solve this more generically are definitely welcomed.

**Comment by mattcui** [comment]
As I mentioned in https://github.com/cloudfoundry/bosh-agent/issues/184#issuecomment-414048419, I think we could reuse [ScrubEphemeralDisk](https://github.com/cloudfoundry/bosh-agent/blob/master/platform/linux_platform.go#L59), if `ScrubEphemeralDisk=true` set in `agent.json` (for Softlayer), the agent will check ephemeral_disk_partitioned property (maybe in `agent_settings.json`) and go through ephemeral disk partition logic, for other infrastructures, no `ScrubEphemeralDisk` set in `agent.json`, so it doesn't need to go through this part of logic, what do you think? Thanks.

@dpb587-pivotal @jfmyers9

**Comment by luan** [comment]
@edwardstudy is this resolved now that #190 was merged?

**Comment by jaresty** [comment]
Closing due to inactivity; please feel free to reopen if you think this is still an issue.
